ldcpy/util.py

```diff
@@ -324,98 +259,129 @@ def check_metrics(
+    June 2019, pp. 517-528 (doi:10.1111/cgf.13707).
+
+    Default tolerances for the tests are:
+    ------------------------
+    K-S: fail if p-value < .05 (significance level)
+    Pearson correlation coefficient:  fail if coefficient < .99999
+    Spatial relative error: fail if > 5% of grid points fail relative error
+    SSIM: fail if SSIM < .99995
 
-    print('Evaluating 4 metrics for {} data (set1) and {} data (set2)'.format(set1, set2), ':')
+    """
 
+    print(f'Evaluating 4 metrics for {set1} data (set1) and {set2} data (set2):')
+    aggregate_dims = metrics_kwargs.pop('aggregate_dims', None)
     diff_metrics = DiffMetrics(
-        ds[varname].sel(collection=set1).isel(time=time),
-        ds[varname].sel(collection=set2).isel(time=time),
-        ['lat', 'lon'],
+        ds[varname].sel(collection=set1),
+        ds[varname].sel(collection=set2),
+        aggregate_dims,
+        **metrics_kwargs,
     )
 
+    # count the number of failures
+    num_fail = 0
     # Pearson less than pcc_tol means fail
     pcc = diff_metrics.get_diff_metric('pearson_correlation_coefficient').data.compute()
     if pcc < pcc_tol:
         print('     *FAILED pearson correlation coefficient test...(pcc = {0:.5f}'.format(pcc), ')')
         num_fail = num_fail + 1
     else:
         print('     PASSED pearson correlation coefficient test...(pcc = {0:.5f}'.format(pcc), ')')
-
     # K-S p-value less than ks_tol means fail (can reject null hypo)
     ks = diff_metrics.get_diff_metric('ks_p_value')
     if ks < ks_tol:
         print('     *FAILED ks test...(ks p_val = {0:.4f}'.format(ks), ')')
         num_fail = num_fail + 1
     else:
         print('     PASSED ks test...(ks p_val = {0:.4f}'.format(ks), ')')
-
     # Spatial rel error fails if more than spre_tol
     spre = diff_metrics.get_diff_metric('spatial_rel_error')
     if spre > spre_tol:
         print('     *FAILED spatial relative error test ... (spre = {0:.2f}'.format(spre), ' %)')
         num_fail = num_fail + 1
     else:
         print('     PASSED spatial relative error test ...(spre = {0:.2f}'.format(spre), ' %)')
-
     # SSIM less than of ssim_tol is failing
     ssim_val = diff_metrics.get_diff_metric('ssim')
     if ssim_val < ssim_tol:
         print('     *FAILED SSIM test ... (ssim = {0:.5f}'.format(ssim_val), ')')
         num_fail = num_fail + 1
     else:
         print('     PASSED SSIM test ... (ssim = {0:.5f}'.format(ssim_val), ')')
-
     if num_fail > 0:
-        print('WARNING: {} of 4 tests failed.'.format(num_fail))
-
+        print(f'WARNING: {num_fail} of 4 tests failed.')
     return num_fail
 
 
-def subset_data(ds, subset=None, lat=None, lon=None, lev=None, start=None, end=None):
+def subset_data(
+    ds,
+    subset=None,
+    lat=None,
+    lon=None,
+    lev=None,
+    start=None,
+    end=None,
+    time_dim_name='time',
+    vertical_dim_name='lev',
+    lat_dim_name='lat',
+    lon_dim_name='lon',
+):
```

Comment:
I tried my best to generalize this function. However, I do think that in its current state, the implementation is fragile and may not generalize well in the future....

Reply:
We are not covering all cases here to guarantee that the user won't have to do the subsetting themselves via xarray. For example, consider a dataset with dimensions `(member_id, time, lev, lat, lon)`, the function in its current state can't subset along the `member_id` dimension.

Reply:
True, I guess my goal was to capture the more common ways to subset the data, but there is probably no way it would cover all cases. I will leave it as a convenience function for now for those who might want to use it... If we rewrite the plot code to accept a single metrics object then users could use subset_data or do their own subsetting before plotting.  The only other place it is used in the code is in check_metrics and compare_stats, but it is only used to select a single level there so I don't think that would need to change.